subsys/storage/flash_map/flash_map.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2015 Runtime Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <device.h>
+#include <flash_map.h>
+#include <flash.h>
+#include <soc.h>
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+struct layout_data {
+	u32_t area_idx;
+	u32_t area_off;
+	u32_t area_len;
+	void *ret;        /* struct flash_area* or struct flash_sector* */
+	u32_t ret_idx;
+	u32_t ret_len;
+	int status;
+};
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+struct driver_map_entry {
+	u8_t id;
+	const char * const name;
```

Comment:
When implementing multiple flash support [MCUB-75](https://runtimeco.atlassian.net/projects/MCUB/issues/MCUB-75) for mcuboot,  I went a similar approach as the `driver_map_entry`, but included the a pointer to `struct device` itself.
It helps to avoid many calls to `device_get_binding(..)`. Instead, it is called once and cached.

Could this be an option?